---
title: "Bulk RNA sequencing"
author: "Mohamed Nadhir Djekidel"
date: "May 11, 2018"
output: html_document
---

# Introduction




# load libraries 

```{r}
require(DESeq2)
require(edgeR)
require(ggplot2)
require(gridExtra)
require(eulerr)
require(VennDiagram)

rootDir <- "D:/Projects/Dux_project/Data/BulkRNASeq"


sampleNames <- list.dirs(rootDir,recursive = F,full.names = F)
sampleNames <- grep("Rep",sampleNames,value = T)
sampleNames
```




# Get Gene expression 

For each sampe we have two directories 'genes' and 'repeat' that each containts 'ReadsPerGene.out.tab' file generated by STAR.
```{r}

genes_expr_files <- list()


for(smp in sampleNames){
  
  smp_dir <- file.path(rootDir,smp,"genes")  
  genes_expr_files[[smp]] <- list.files(path = smp_dir,pattern = "*ReadsPerGene.out.tab")
  
  genes_expr_files[[smp]] <- file.path(smp_dir,genes_expr_files[[smp]])
}

```




# Get repeat expression

```{r}


repeats_expr_files <- list()


for(smp in sampleNames){
  
  smp_dir <- file.path(rootDir,smp,"repeats")  
  repeats_expr_files[[smp]] <- list.files(path = smp_dir,pattern = "*ReadsPerGene.out.tab")
  
  repeats_expr_files[[smp]] <- file.path(smp_dir,repeats_expr_files[[smp]])
}

```



```{r}
require(rtracklayer)
mm10_geneNames <- import.gff("D:/genomes/GRCm38/annotation/Mus_musculus.GRCm38.85_fixed.gtf")
mm10_geneNames <- subset(mm10_geneNames, type=="gene")

ens2geneName <- as.data.frame(mcols(mm10_geneNames)[,c("gene_id","gene_name")])
head(ens2geneName)
```

```{r}
require(rtracklayer)
mm9_geneNames <- import.gff("D:/genomes/mm9/Mus_musculus.NCBIM37.67_fixed.gtf")

ens2geneName_mm9 <- as.data.frame(mcols(mm9_geneNames)[,c("gene_id","gene_name")])

ens2geneName_mm9 <- unique(ens2geneName_mm9)

head(ens2geneName_mm9)
```

# Merge genes and repeats expression

```{r}

exprs_all <- matrix(0,nrow = 0,ncol = length(sampleNames))

colnames(exprs_all) <- sampleNames


for(i in 1:length(sampleNames)){

  
  smp =sampleNames[i]
  
  print(smp)
  
  genes_expr <- read.table(genes_expr_files[[smp]],skip = 4,stringsAsFactors = F)
  
  #pos <- match(genes_expr$V1,ens2geneName$gene_id)
  
  #toUse <- which(!is.na(pos))
  #genes_expr$V1[toUse] <- ens2geneName$gene_name[pos[toUse]]
  
  repeats_expr <- read.table(repeats_expr_files[[smp]],skip = 4,stringsAsFactors = F)
  
  if(i == 1){
    gnames <- c(genes_expr[,1], repeats_expr[,1])
    
    exprs_all <- matrix(0,nrow = length(gnames),ncol = length(sampleNames))
    
    colnames(exprs_all) <- sampleNames
    rownames(exprs_all) <- gnames
    
  }
  
  # Write genes expression
  exprs_all[genes_expr[,1],smp] <- genes_expr[,2]
  
  # write repeats expression
  exprs_all[repeats_expr[,1],smp] <- repeats_expr[,2]  
  
}


## Remove genes with 0 counts in all samples

#nbExpr <- apply(exprs_all, 1, function(x) sum(x>0))
#exprs_all <- exprs_all[nbExpr >0,]

head(exprs_all)
```



# Prepare Meta data 


```{r}

samples.metaData <- data.frame(row.names = colnames(exprs_all),
                               SampleNames = colnames(exprs_all),
                               type = factor(gsub("_Rep[12]","",colnames(exprs_all))) )
samples.metaData
```


# Create edgeR object

```{r}

allSmp.edgeRObj <- DGEList(exprs_all,
                           samples = samples.metaData,
                           group = samples.metaData$type)
allSmp.edgeRObj$counts[1:10,]
```


# Normalize Data

```{r}
allSmp.edgeRObj <- calcNormFactors(allSmp.edgeRObj,method = "TMM")
cpm_norm <- log2(cpm(allSmp.edgeRObj,normalized.lib.sizes = T)+0.1)


for(grp in levels(samples.metaData$type)){
  
  
  smps <- grep(grp,colnames(cpm_norm),value = T)
  
  ttl <- paste(smps[1],"vS",smps[2])
  
  toUse <- apply(cpm_norm[,smps], 1, function(x) sum(x>1))
  
  smoothScatter(cpm_norm[toUse>0,smps],main=ttl)
  abline(0,1,col="red")
  
  tmp_df <- as.data.frame(cpm_norm[,smps])
  
  colnames(tmp_df) <- c("x","y")
  rs <- lm(x~y,data = tmp_df)
  
  r2 <-  summary(rs)
  
  text(10,1,paste0("r=",sprintf("%.2f",r2$adj.r.squared)),cex = 1)
}


```


```{r fig.height=4, fig.width=6}
require(RColorBrewer)
require(ggsci)
require(ggrepel)


pca.data <- prcomp(t(cpm_norm),retx = TRUE,center = FALSE, scale. = FALSE)


pca.data_df <- data.frame(pca.data$x,
                          Sample = factor(samples.metaData$SampleNames),
                          treatment= factor(samples.metaData$type))

ggplot(pca.data_df,aes(x=PC1,y=PC2,col=treatment,label=Sample))+
   geom_point(size=3)+ 
   scale_color_brewer(palette = "Set1") +
   xlab("PC1") + 
  ylab("PC2") +
  # xlab(paste("PC1 (", percent[1], "%)", sep = "")) + 
  #ylab(paste("PC2 (", percent[2], "%)", sep = "")) +
   theme_bw() + theme(text = element_text(face="bold",color = "black"),
                      axis.text = element_text(face="bold",color = "black"),
                      plot.title = element_text(hjust = 0.5)
                      ) +
  ggtitle("PCA samples") #+  geom_text_repel(color="black")

```


## Get the differentially expressed genes


```{r}

designMat <- model.matrix(~0+type, data = samples.metaData)
colnames(designMat) <- levels(samples.metaData$type)

allSmp.edgeRObj <- estimateGLMCommonDisp(allSmp.edgeRObj, design=designMat)
#allSmp.edgeRObj <- estimateGLMTagwiseDisp(allSmp.edgeRObj, design=designMat)
#allSmp.edgeRObj <- estimateGLMTrendedDisp(allSmp.edgeRObj, design=designMat)


#allSmp.edgeRObj <- estimateGLMRobustDisp(allSmp.edgeRObj, designMat)

fit <- glmQLFit(allSmp.edgeRObj, designMat)


Comparisons <- list()

Comparisons[["D0_Vs_D1Neg"]] <- c(-1,1,0)
Comparisons[["D0_Vs_D1Pos"]] <- c(-1,0,1)
Comparisons[["D1Pos_Vs_D1Neg"]] <- c(0,-1,1)


DEGresults <- list()



for(cmp in names(Comparisons)){
  print(cmp)
  
  DEGresults[[cmp]] <- glmQLFTest(fit, contrast=Comparisons[[cmp]])
}


```


## Save results

```{r}
pval_thr = 1e-3
fc_thr=log2(2)


dir.create("DEGresults",showWarnings = FALSE)
for(cmp in names(DEGresults)){

  print(cmp)
  fout = paste0(cmp,".csv")
  fout = file.path("DEGresults",fout)
  
  res <- as.data.frame(topTags(DEGresults[[cmp]], n = nrow(exprs_all)))
  
  res$isDE = "NO"
  
  pos = which(res$logFC > fc_thr & res$FDR < pval_thr)
  res$isDE[pos] <- "Up"
  
  pos = which(res$logFC < -fc_thr & res$FDR < pval_thr)
  res$isDE[pos] <- "Down"
  
  ## Add gene name
  pos <- match(rownames(res),mm10_geneNames$gene_id)
  
  notNA <- which(!is.na(pos))
  
  res$geneName <- rownames(res)
  res$geneName[notNA] <- mm10_geneNames$gene_name[pos[notNA]]
  
  write.csv(res,file = fout)
}

```


## Create heatmap of all the DE genes 

```{r fig.height=10, fig.width=3}
pval_thr = 1e-3
fc_thr=log2(2)

allDEgenes <- c()

for(cmp in names(DEGresults)){

  print(cmp)
  fout = paste0(cmp,".pdf")
  fout = file.path("DEGresults",fout)
  
  res <- as.data.frame(topTags(DEGresults[[cmp]], n = nrow(allSmp.edgeRObj)))
  deg = rownames(subset(res, abs(logFC) > fc_thr & FDR < pval_thr ))
  
  
  allDEgenes <- c(allDEgenes,deg)

}

allDEgenes <- unique(allDEgenes)


cols = colorRampPalette(rev(brewer.pal(n = 7, name ="RdYlBu")))(100)


expr_scaled <- pheatmap:::scale_rows(norm_expr[allDEgenes,])


#expr_scaled[expr_scaled>2]=2
#expr_scaled[expr_scaled < -2]= -2


pos <- match(c("MERVL-int","Zscan4a","Zscan4c","Zscan4d","MT2_Mm","Glul","Zfp36l1","Btg2","Smad7"), rownames(expr_scaled))

H=Heatmap(expr_scaled,show_column_dend = F,show_row_dend = F,show_row_names = F,col = cols,
          split = 8,cluster_columns = F,
        name = "Relative expression") +
  rowAnnotation(link = row_anno_link(at = pos, 
                                     labels = c("MERVL-int","Zscan4a","Zscan4c","Zscan4d","MT2_Mm","Glul","Zfp36l1","Btg2","Smad7")),
                width = unit(0.1, "cm") + max_text_width(labels))

draw(H)

for (cls in 1:8) {
        decorate_heatmap_body("Relative expression", {
          grid.rect(gp = gpar(col = "black", fill = NA))
        }, slice = cls)
}

clus <- row_order(H)

```


## Plot heatmap of the DE genes between D0 and D1 

```{r fig.height=4, fig.width=4}
pval_thr = 1e-3
fc_thr=log2(2)

D0_Vs_D1Pos_DEG <- as.data.frame(topTags(DEGresults[["D0_Vs_D1Pos"]], n = nrow(allSmp.edgeRObj)))

D0_Vs_D1Pos_DEG$isDE = "NO"
  
pos = which(D0_Vs_D1Pos_DEG$logFC > fc_thr & D0_Vs_D1Pos_DEG$FDR < pval_thr)
D0_Vs_D1Pos_DEG$isDE[pos] <- "Up"
  
pos = which(D0_Vs_D1Pos_DEG$logFC < -fc_thr & D0_Vs_D1Pos_DEG$FDR < pval_thr)
D0_Vs_D1Pos_DEG$isDE[pos] <- "Down"


D1POS_UPGENES <- rownames(subset(D0_Vs_D1Pos_DEG, isDE=="Up"))

D1POS_UPGENES_mean = matrix(0,nrow = length(D1POS_UPGENES),ncol = 3)
rownames(D1POS_UPGENES_mean) = D1POS_UPGENES

D1POS_UPGENES_mean[,1] = rowMeans(cpm_norm[D1POS_UPGENES,c("D0WT_Rep1","D0WT_Rep2")])
D1POS_UPGENES_mean[,2] = rowMeans(cpm_norm[D1POS_UPGENES,c("D1_2CNeg_Rep1","D1_2CNeg_Rep2")])
D1POS_UPGENES_mean[,3] = rowMeans(cpm_norm[D1POS_UPGENES,c("D1_2CPos_Rep1","D1_2CPos_Rep2")])

colnames(D1POS_UPGENES_mean) <- c("D0WT","D1_2CNeg","D1_2CPos")

D1POS_UPGENES_mlt <- melt(D1POS_UPGENES_mean)


my_comparison <- list(c("D0WT","D1_2CNeg"), c("D1_2CNeg","D1_2CPos"))

ggplot(D1POS_UPGENES_mlt, aes(x=Var2,y=value,fill=Var2)) + geom_boxplot() + theme_bw() + scale_fill_nejm() +
   stat_compare_means(comparisons = my_comparison,hide.ns = T,method = "wilcox.test",paired = FALSE)
```

